driver/sdk.md

```diff
@@ -76,78 +84,269 @@ $ bblfsh-sdk update
+The core functionality to be implemented on the parser side is the source code
+parser.  This should get the source contents of a file and return an AST
+representation.  This AST representation should be what the native or library AST
+parser outputs as is. The native parser might be from the standard library of the
+language, a third-party library, or a custom one (in that order of preference).
+Any language can be used to implement the parser so use whatever is easier for the
+specific language you're working on.
 
-### JSON serialization
+#### Native parser operation
 
-The driver should contain a JSON serializer which is able to serialize requests
-and responses (including the AST) into a single-line JSON.
+When the driver executes the parser, the former entry point should Officially a
+main loop similar to the example below.  It should read requests from standard
+input (which contain a string with the contents of the source to parse), parse it
+generating the native AST and write a response to the standard output.  
 
-### Main loop
+Requests and responses must be serialized in JSON format on a single line finished
+with a Unix newline character ('\n') so the driver can iterate over the standard
+input line by line.  This means so the language selected to implement the parser
+should support it in the standard library or any third party module. 
 
-When the parser is executed, its entry point should execute this main loop.
-It should read requests from standard input, process them and write responses to
-standard output. A response for a request should always be written, even if
-processing fails.
+A response for a request should always be written, even if processing fails (in
+this case an error response will be returned). If the standard input stream is
+closed or the program can write to it for any other reason, the program should
+exit with code 0.  If the writing to standard output fails, the program should
+exit with code 1 (additionally it could try to log the error to stderr).  
 
-If the standard input stream is closed, the program should exit with code 0.
-If write to standard output fails, the program should exit with code 1.
-Here we illustrate its behaviour in Python-like syntax:
+Here we illustrate its behaviour in Python-like pseudocode:
 
 ```python
 while True:
     line = read_line_from_stdin()
     if stdin_is_closed():
         exit(0)
+
     try:
         request = parse_json_request(line)
-        ast = parse_ast(request)
+        ast = parse_ast(request["content"])
         response = serialize_json_response(ast)
-    except:
+    except Exception, e:
         try:
-            write_to_stdout(fatal_response)
+            write_to_stdout(generate_error_response(e))
         except:
             exit(1)
         continue
+
     try:
         write_to_stdout(response)
     except:
         exit(1)
 ```
 
-## Normalizing the AST
+In order to avoid maintainability burdens, the parser should avoid as much as
```

Comment:
Native speaker: "when the sentence is long like in this case it sounds better before than after, but add commas" (added).